### ISSUE TITLE
Scrollbar resizing and jumping

### ISSUE THREAD

**Issue by VadymPinchuk** [flutter/flutter#25652]
Hi everyone. 
I have noticed that scrollbar working fine with Lists without pagination and with items of the same size.
While we have pagination and items are of different size (like chat with many messages) - scrollbar behaves strange - it is jumping and resizing itself

You can find reference to video here:
https://drive.google.com/open?id=1sCmo5_22juM45q0kT82nQdvPwzCwJ-zU

Please let me know what can I do, or, if there any solutions for that?
Thank you in advance

**Comment by DoctorWho-01** [comment]
Hi @zoechi - could you please provide some feedback on this as we need to see if we need a work around if you cannot provide a solution.

**Comment by marcglasberg** [comment]
I would like to present some ideas as how to fix the Scrollbar behavior. I suggest a few new parameters:

1) minSize : the minimum size the Scollbar can assume.

2) maxSize : the maximum size the Scollbar can assume.

3) sizeBy: the current behavior is sizeBy=SizeBy.pixels. However, if sizeBy=SizeBy.itemCount, the bar position and size would relate to the index of the top and bottom items being shown in the screen, not their sizes.

These 3 parameters may be carefully chosen by the developer who knows the kind of content their scrollable will have, as to reduce resizing and jumping. Obviously it can never be perfect, since not always it's possible to know where you are in the list, the size of all items etc.

While at that, I would suggest a 4th parameter:

4) visible: forces the scrollbar to be visible at all times.

**Comment by layou1989** [comment]
Has this problem been solved?

**Comment by VladyslavBondarenko** [comment]
Can be reproduced with licenses page at https://flutter.github.io/gallery/
https://github.com/flutter/flutter/issues/15967#issue-308912264

**Comment by miguelpruivo** [comment]
This problem is terrible, specially in chat screens. Any workarounds for it?

**Comment by ajviera** [comment]
Does anyone have a solution for this? It is very annoying and several customers complain about the behavior of the ScrollBar

**Comment by ilaurillard** [comment]
While wrapping a ListView into a SingleChildScrollView fixes Scrollbar twitching for me .. i do pay with smoothness of scrolling for that. After scrolling down about 100 (dynamic height) items .. perfomance gets low and jerky .. :/

Hope for solution!!

**Comment by VladyslavBondarenko** [comment]
Since license page is changed, tried with this code sample

```
import 'dart:math';

import 'package:flutter/material.dart';

void main() => runApp(MaterialApp(home: MyHomePage()));

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  var rng = Random();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Scrollbar(
        child: ListView.builder(
          itemBuilder: (BuildContext context, int index) {
            String text = "Item at $index";
            if (index % 5 == 0) {
              text += ". Very long label that should wrap. " * rng.nextInt(10);
            }
            return Padding(
              padding: EdgeInsets.all(10.0),
              child: Text(text),
            );
          },
          itemCount: 5000,
        ),
      ),
    );
  }
}
```

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/17024977/88669215-495e5680-d0ec-11ea-9c3c-eb9bdabdf3af.gif)

Persists with current master `1.21.0-6.0.pre.41`

<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.21.0-6.0.pre.41, on Mac OS X 10.15.6 19G73, locale en-GB)
    • Flutter version 1.21.0-6.0.pre.41 at /Users/nevercode/dev/flutter
    • Framework revision 39fa00201d (3 hours ago), 2020-07-27 21:11:43 -0700
    • Engine revision f27729e97b
    • Dart version 2.10.0 (build 2.10.0-0.0.dev 24c7666def)
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

 
[✓] Connected device (5 available)            
    • Redmi Note 7 (mobile)       • 6345c469                                 • android-arm64  • Android 9 (API 28)
    • Nevercode’s iPhone (mobile) • b668e524315069f3db3661ac11ff1f66afafebdb • ios            • iOS 13.6
    • macOS (desktop)             • macos                                    • darwin-x64     • Mac OS X 10.15.6 19G73
    • Web Server (web)            • web-server                               • web-javascript • Flutter Tools
    • Chrome (web)                • chrome                                   • web-javascript • Google Chrome 84.0.4147.105

• No issues found!
```

</details>

**Comment by ozz-rjq** [comment]
The same issue when using `ListView.builder`, is it possible to fix?

**Comment by wwwdata** [comment]
It has been a while now since this is an issue. Is this actively worked on? This really prevents me from using a scrollbar altogether if my ListView content is dynamic and not each element is of the same height.

**Comment by devj3ns** [comment]
Is there still no fix for this?

**Comment by MaximMonakhov** [comment]
Ran into the same weird thing using a dynamic list. I had to remove the scrollbar. It is clear that if the sizes in the list are different, it is probably difficult to calculate the position of the scrollbar. But at least let it not jump, but do it smoothly, with animation

**Comment by MaximMonakhov** [comment]
Ran into the same weird thing using a dynamic list. I had to remove the scrollbar. It is clear that if the sizes in the list are different, it is probably difficult to calculate the position of the scrollbar. But at least let it not jump, but do it smoothly, with animation

**Comment by gcostaapps** [comment]
I'm facing the same issue, any workaround?

**Comment by KartikShrivastava** [comment]
Scroll bar is also re-sizing and shifting position while scrolling for me. Does anyone found a fix or workaround?

**Comment by son** [comment]
```
Scrollbar(
  controller: scrollController, 
  child: ListView(
    cacheExtent: 10000, // higher than content
  ),
);
```

**Comment by Piinks** [comment]
This behavior is due to the fact that the scrollbar represents the relative position of the scroll view over the full length of the scroll view. If you have dynamically sized items that are being built lazily, there is no way to reliably know ahead of time how long the scroll view is (the maxScrollExtent) - until everything is built.

So what you are seeing as you scroll is that the scrollbar is updating with more relevant information as items are being lazily built and the maxScrollExtent is updated, because it does not know that information ahead of time.

I thought of this while working on https://github.com/flutter/flutter/issues/41434
That issue is about scrollbar in infinitely long scroll views. Im putting together a proposal for scrollbar behavior when the maxScrollExtent is infinity, would it sound appropriate to folks here that this kind of feature could also be extended to cover this use case?

Basically, the scrollbar would operate on an assumption, pretending to know how long the scroll view is, maintaining smooth animations and gesture support. This is common in apps where you see the scrollbar update after it reaches the bottom and loads more content.

If that sounds like something that would resolve this issue, let me know, and I will incorporate it into my proposal. Thanks all!

**Comment by ilaurillard** [comment]
Mayhaps it would be possible to assume/calculate an average item-height (based on built items so far) and set maxExtent calculation dynamically?

**Comment by Piinks** [comment]
> Mayhaps it would be possible to assume/calculate an average item-height (based on built items so far) and set maxExtent calculation dynamically?

This is actually how it works currently, it tries to make an estimate, and updates that estimate as it gets more information.

**Comment by ertong** [comment]
>Mayhaps it would be possible to 
>assume/calculate an average item-height

I think, this approach is dead end.
We cannot know item sizes and cannot assume them.

But we can know item count, index of the first item on the screen, index of the last item on the screen and what part of the first item and the last item is visible.

So, we know how much items before the screen, on the screen and after the screen. It looks like it is enough to show smooth scrollbar (if we count partially showed items as fraction).

As far as I understand, this approach is default in native android recycled view.

**Comment by Piinks** [comment]
There is a prototype and design document available here for those interested in addressing this issue per [my comment here.](https://github.com/flutter/flutter/issues/25652#issuecomment-990470625) Feedback is welcome. :)

[Design Document](https://docs.google.com/document/d/1TV2oF2iLIZtGd48336korxMTQrGizDAASAeZAeteMSc/edit?usp=sharing)

PR: https://github.com/flutter/flutter/pull/96825

**Comment by Leffe108** [comment]
While there it is not always the case that we know the extent of the children, there are some cases in which it would be useful if itemExtent and/or prototypeItem is a function of item index. Which I understand have implications on performance as you have to call it for each item. And ideally prototypeItem would somehow be a pointer so if the list have many items that uses the same prototypeItem it only has to be layouted once. Maybe it can be archived with keys on the prototype item? 

In the issue I mentioned above, there is a reproduction case where it would be possible to provide prototypeItem as a function of item index and in that case have a perfectly functioning scrollbar. itemExtent would also be possible if Flutter exposed ListTile height as some sort of constant.

**Comment by maheshmnj** [comment]
I can reproduce this issue on both the stable and the master channel. @Piinks as per the below code sample the issue can be reproduced even when we know the listItems beforehand (no building lazily).

*Notice how the scrollbar jumps in the below video*

https://user-images.githubusercontent.com/31410839/151351589-c57c625e-8ac0-4ff9-89c6-bc0ee9450769.mov


<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.8.1, on macOS 12.1 21C52 darwin-arm, locale en-GB)
    • Flutter version 2.8.1 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (3 weeks ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 60.1.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.63.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.32.0

[✓] Connected device (4 available)
    • Redmi K20 Pro (mobile) • 192.168.1.3:5555                     • android-arm64  • Android 11 (API 30)
    • iPhone 13 Pro (mobile) • 0FA2FB93-2A89-4FDB-8293-DF66188F1E54 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-0
      (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 12.1 21C52 darwin-arm
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 96.0.4664.110

• No issues found!
```

```
[✓] Flutter (Channel master, 2.10.0-1.0.pre.307, on macOS 12.1 21C52 darwin-arm, locale en-GB)
    • Flutter version 2.10.0-1.0.pre.307 at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b7424c619c (7 hours ago), 2022-01-26 17:25:45 -0800
    • Engine revision c248c6e38e
    • Dart version 2.17.0 (build 2.17.0-63.0.dev)
    • DevTools version 2.10.0-dev.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 60.1.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.63.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.32.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 12.1 21C52 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 97.0.4692.99

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>


<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  Widget box(double height) {
    return Padding(
      padding: const EdgeInsets.all(16.0),
      child: Column(
        children: [
          Text("test $height"),
          SizedBox(
            height: height,
            width: 30,
            child: const DecoratedBox(
              decoration: const BoxDecoration(color: Colors.red),
            ),
          ),
        ],
      ),
    );
  }

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Jumpy Scrollbar',
        home: Scaffold(
          appBar: AppBar(title: const Text('Jumpy Scrollbar')),
          body: Scrollbar(
            isAlwaysShown: true,
            child: ListView(children: [
              box(8000),
              box(200),
              box(200),
              box(200),
            ]),
          ),
        ));
  }
}

```
</details>

**Comment by chipweinberger** [comment]
I’m trying to build a slider that works like the Google Photos app and lets you drag around through *years* of data.

This scrolling behavior is a showstopper for me and this is my first week of using Flutter. It’s crazy this issue is so old. 

In my use case:
- i know ahead of time how much content there is (even if it’s not all loaded)
- i know ahead of time how much content will be loaded when a user takes a new photo
- if I’m soothly increasing those item heights i set, then the scroll bar should be smoothly moving into its new calculated location

But as @maheshmnj shows, it also reproduces on static lists of static heights that are all rendered at once! 

That’s crazy to me.

I’d love a solution that:
-  let’s me tell it the sizes of all the items in the list ahead of time
-  I can update those item sizes if I change them or know they will change. 
- if item sizes change

**At worst, you’d at least expect default behavior of the ListView to remember sizes of children it renders and to restore to smooth scroll bar behavior after the initial jump.** But the scroll bar remains jumpy even after all items have been rendered.

 This needs to be fixed!

**Comment by Piinks** [comment]
Hi @maheshmnj your reproduction illustrates the same issue. The estimated max extent is based on the first child in your case, which is much much larger than the rest. So if you check the ScrollMetrics after building your app, they are around 32,000 (depending on your window size), which reflects the first one being used as a prototype to estimate the full extent. (8000 * 4)

The scrollbar jumps when you scroll down because the max extent is updated (to around 8600) once those other (smaller) children come into view. In your case it is still estimating the extent.

I wonder if something like a prototypeExtent would help... prototypeItem forces all of the children to be of the same size, but maybe we can allow for developers to provide their own estimate, one that would not force the children to be a given size. I feel like it could help resolve these issues, but I don't think it is a very good experience.

No matter how perfect we might be able to estimate in one situation, there will always be another case where that process of estimation is way off.
Maybe then the right solution is to figure out a couple of good, consistent patterns of making this estimate, and exposing the estimation choice to the developer. Not sure, but interested to hear what folks think.

**Comment by Piinks** [comment]
Hi @chipweinberger

>  - let’s me tell it the sizes of all the items in the list ahead of time
> - I can update those item sizes if I change them or know they are changing.
> - The scroll bar only jumps once initially on item size changes, or I can control how quickly it moves to its new location

This is all already possible. :)

From your comments it sounds like you are new to Flutter - welcome! If the items have a static extent, have you tried using ListView.itemExtent? Without a reproduction of your issue, it is hard to understand the behavior you are seeing. 

I would recommend filing a new issue with a reproduction so we can diagnose it. If it is the same issue, we can close it and link it back to this issue.

First though, please do read our code of conduct: https://github.com/flutter/flutter/blob/master/CODE_OF_CONDUCT.md

**Comment by chipweinberger** [comment]
Thanks Piinks!

I added a few reproducing cases in #97354 with the code. 

In my use case, the first item in the list can expand depending on data received over bluetooth. I will give itemExtent a shot and see if that can fix it.

**Comment by Leffe108** [comment]
Would it be possible to provide an API for itemExtent as a function of item index? 

[There is a case here](https://github.com/flutter/flutter/issues/80445) where it would be possible for the code to calculate the height of each item as long as you know how tall a ListTile is (you could probably get away with hard coding that into the app and then research ways to obtain that height dynamically as a function of UI density, ultra-large font size etc.)

**Comment by KaelTeck** [comment]
`
Scrollbar(
     child:  MasonryGridView.count(
           key: _scrollViewKey,
           crossAxisCount: 4,
           mainAxisSpacing: 5.0,
           crossAxisSpacing: 5.0,
           padding: _padding,
           physics: _scrollPhysics,
           scrollDirection: _scrollDirection,
           shrinkWrap: false, // Disabling shrinkWrap
           itemBuilder: itemBuilder,
           itemCount: getItemCount(),
           cacheExtent: 2500, // Adding cacheExtend
           reverse: _reverse,
           primary: true,
           addAutomaticKeepAlives: _addAutomaticKeepAlives,
           addRepaintBoundaries: _addRepaintBoundaries,
        ),
);`
That fixed it for me.

**Comment by au-top** [comment]
Maybe we don't need to know the height, just index and count? At least the following code helps me alleviate this problem in my chat app, but it relies on [listview.build] to monitor the loaded index value. I think this is a simple method, just an idea

```

class FreeScrollBar extends StatefulWidget {
  final Widget child;
  final double barWidth;
  final double barHeight;
  const FreeScrollBar({
    super.key,
    required this.child,
    required this.barHeight,
    required this.barWidth,
  });

  @override
  State<FreeScrollBar> createState() => FreeScrollBarState();

  static FreeScrollBarState of(BuildContext context) {
    return context.findAncestorStateOfType<FreeScrollBarState>()!;
  }
}

class FreeScrollBarState extends State<FreeScrollBar> with ThemeStyleMixin {
  final _nowIndex = ValueNotifier(0);
  final _count = ValueNotifier(1);

  int lateIndex = 0;

  void setCount(int count) {
    _count.value = count;
  }

  void setNowIndex(int index) {
    _nowIndex.value = index.clamp(0, _count.value);
  }

  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        Expanded(child: widget.child),
        LayoutBuilder(
          builder: (context, boxConstraints) => Container(
            color: colorScheme.surface,
            width: 8,
            height: double.infinity,
            child: AnimatedBuilder(
              builder: (context, childBlock) => Stack(
                children: [
                  AnimatedPositioned(
                    duration: kThemeAnimationDuration,
                    top: (boxConstraints.maxHeight * (1 - _nowIndex.value / _count.value)).clamp(
                      0,
                      boxConstraints.maxHeight - widget.barHeight,
                    ),
                    child: childBlock!,
                  ),
                ],
              ),
              animation: Listenable.merge([_nowIndex, _count]),
              child: Container(
                height: widget.barHeight,
                width: widget.barWidth,
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(10),
                  color: colorScheme.onSurface,
                ),
              ),
            ),
          ),
        ),
      ],
    );
  }
}


```

**Comment by Leffe108** [comment]
@au-top 
> Maybe we don't need to know the height, just index and count? At least the following code helps me alleviate this problem in my chat app, but it relies on [listview.build] to monitor the loaded index value. I think this is a simple method, just an idea
> 

Your example seem to demonstrate an example where all rows are of equal height. Your proposed solution may work if that holds, but what if there are one row that is 10x taller than the others? Perhaps you have a list of expansion pannels that can be opened.

**Comment by au-top** [comment]
Sorry, I just solve its weird jumping behavior. This is a compromise mitigation solution, not a solution. As seen in the video, I rely on the builditem index parameter of listview to control the position of the scroll bar, so it has nothing to do with the sub item widget height However, if the build item progress bar does not occur, it will be disabled I think this can alleviate some visual problems before the fluent repair


https://user-images.githubusercontent.com/44134842/179939379-3791f64d-d795-465f-96af-2a217273ccfe.mp4

```

int count=10;

ListView.build(

 itemCount: count,
...
 itemBuilder: (context, index) {

        Future(() async {
            FreeScrollBar.of(context).setCount(count);
            FreeScrollBar.of(context).setNowIndex(index);
          });

         return  YouWidget() ; 
     }
...
)

```

**Comment by waslost0** [comment]
Any news?

**Comment by megur0** [comment]
> This is an example that does that: https://gist.github.com/kekland/7d0e1257957a3bf84b4f033377635eb0

Inspired by kekland, I created a program that precomputes the size in ListView. The sample case is like a chat message with text and images.
https://gist.github.com/megur0/97d1d1e1c1838cede94a911d9fce5978

In this program, the render tree is pre-created with the Element.mount() method and its size is calculated with the computeDryLayout() method.
This way is not a common way, so unexpected errors may occur, but it seems to cover some use cases.
```
// Create dummy BuildOwner.
BuildOwner owner = BuildOwner(focusManager: FocusManager());

// Create dummy render tree with inner widgets.
// Using lockState, buildScope function to avoid assert error.
_WrapElement? element;
owner.lockState(() {
    element = widget.createElement();
    element!.assignOwner(owner);
});
owner.buildScope(element!, () {
    element!.mount(null, null);
});

// Find _WrapColumn widget and get Size
final columnSize = element!
    .findWrapColumnFromDescendant()!
    .computeDryLayout(const BoxConstraints());

// Dispose dummy tree
element!.deactivate();
element!.unmount();
```

**Comment by kekland** [comment]
@megur0 that looks great! Maybe this can be made into a package or something like that. Theoretically things like smoothly scrolling to a specific index in the list is also possible with this as we know the height of every item

**Comment by Piinks** [comment]
I have a proposal for this, but it needs some refinement. I am currently working on 2D scrolling issues as a higher priority, but once resolved I should be able to return to this. I have updated the priority label.

**Comment by VadymPinchuk** [comment]
Happy to give my hand (and time) if I can be useful here as well

**Comment by Piinks** [comment]
Thank you @VadymPinchuk, I'd be happy for your feedback. The proposal is old, but the original design doc and PR were:
- [design document](https://docs.google.com/document/d/1TV2oF2iLIZtGd48336korxMTQrGizDAASAeZAeteMSc/edit?usp=sharing)
- https://github.com/flutter/flutter/pull/96825

Are you on the Discord? Happy to talk more there if you're interested. Someone could always adopt the original PR and pick it up if they wanted to contribute. 
https://github.com/flutter/flutter/wiki/Chat

**Comment by VadymPinchuk** [comment]
It is my #1 priority now @Piinks for tomorrow's morning, after a coffee :)
Will check all the resources first and return back to you

**Comment by JWo1F** [comment]
Hi there! Do you have any progress with creating a library for it?
@VadymPinchuk

**Comment by VadymPinchuk** [comment]
Hi @JWo1F . So far I have some stuff to do :) As soon as I will have a progress you will get an update. Thanks

**Comment by VadymPinchuk** [comment]
@Piinks  I have registered at Discord for this purpose. Please find me: vpinchuk

And lets discuss it there

**Comment by GeylanKalafMohe** [comment]
Kindly let us know if there is any update fixing this issue as it affects all apps using the ListView

**Comment by VadymPinchuk** [comment]
@Piinks I am at Discord. Lets connect there to discuss?

**Comment by Piinks** [comment]
> @Piinks I am at Discord. Lets connect there to discuss?

Sure! Please feel free to reach out in the framework channel! Thanks!

**Comment by VadymPinchuk** [comment]
@Piinks not sure I can get access to it. Pls add me: vpinchuk#4215

**Comment by VadymPinchuk** [comment]
> > @Piinks I am at Discord. Lets connect there to discuss?
> 
> Sure! Please feel free to reach out in the framework channel! Thanks!

Got it. Looking into contributors guideline

**Comment by GeylanKalafMohe** [comment]
any update?????? Please fix this

**Comment by WelliRigo** [comment]
@au-top can you provide more details of your mitigation solution? I found it very interesting, but was not able to reproduce

**Comment by Mik77o** [comment]
This issue still occurs with `ListView` (flutter web).

```
environment:
  sdk: ">=3.1.0 <4.0.0"
  flutter: ">=3.22.1"
```

**Comment by jozefkosecky** [comment]
@Mik77o 
Try this solution. It helped me with my desktop app:

final ScrollController scrollController = ScrollController();

ListView.builder(
        controller: scrollController,
        physics: const AlwaysScrollableScrollPhysics(),
        cacheExtent: 10000,
        itemCount: ....
        itemBuilder: ....
)

If you ask why cacheExtent is 10000, I tried this number and it works

**Comment by Mik77o** [comment]
Yes, checked and it works, thank you!

**Comment by GeylanKalafMohe** [comment]
> @Mik77o 
> 
> Try this solution. It helped me with my desktop app:
> 
> 
> 
> ```
> 
> final ScrollController scrollController = ScrollController();
> 
> 
> 
> ListView.builder(
> 
>         controller: scrollController,
> 
>         physics: const AlwaysScrollableScrollPhysics(),
> 
>         cacheExtent: 10000,
> 
>         itemCount: ....
> 
>         itemBuilder: ....
> 
> )
> 
> ```
> 
> 
> 
> If you ask why cacheExtent is 10000, I tried this number and it works

Yes it works like a charm. Thinking whether it is harmful to apply it to every listview in an app in order to avoid this issue?

**Comment by jozefkosecky** [comment]
@GeylanKalafMohe 
I applied it only where I fetch an 'infinity list' in batches of 5 items, each with different sizes of child widgets. The cacheExtent should pre-compute the size of the next children, so I wouldn't use it everywhere. It can consume a lot of RAM (This is what I think).

**Comment by mattbodley** [comment]
This is also the same kind of problem in the sliverList, which unfortunately doesn't have the cacheExtent quick fix. 

But I think the should be an easy way to solve it at the framework level, since the problem only really happens when someone scrolls down the page (as the items are expanded/contracted downward when they are lazy loaded e.g. with a loading indicator and when once the data is loaded they are rendered with their correct height), all that needs to happen is that when a user scrolls up, the items off screen can expand/contract upwards rather than downwards, so that the end result is the scrolling looks smooth, rather than jumping when they expand/contract to their final size, as otherwise the current approach of pushing (or pulling) the content down/up when scrolling up happens.

This is in the particular case where items are loaded via a future and they are then expanded because the height of each item (row) is not known in advance, e.g. starts with a loading indicator and then expands/contracts to its full height

**Comment by wwwdata** [comment]
In the meantime I can highly recommend this package to anybody that wants a better Scrollbar and other cool features: https://pub.dev/packages/super_sliver_list

**Comment by Nitin-Singh18** [comment]
> ```
> Scrollbar(
>   controller: scrollController, 
>   child: ListView(
>     controller: scrollController,
>     cacheExtent: 10000, // higher than content
>   ),
> );
> ```

Thank you, this solved the problem.

**Comment by jozefkosecky** [comment]
I tried the https://pub.dev/packages/super_sliver_list package as @wwwdata mentioned, and it works very well. I would say it is better than my solution. Scrolling looks smoother, and memory usage is about 100-200 MB less.

**Comment by MathieuPierfitte** [comment]
I've been facing a similar issue and worked on a proof of concept on how to improve this behaviour.

Please have a look at the before & after videos first:

- Before: https://youtube.com/shorts/2pw4UveqRlM?feature=share
- After: https://youtube.com/shorts/WmY3WEehIYM?feature=share (The small bump as I get close to the bottom is probably because my maths in `_computeMaxScrollExtent` is a bit off, see the rest of the post)

Here is a simplified version of my widget tree:

```
Scrollbar(
  child: ListView.builder(
    itemCount: 100,
    itemBuilder: (context, index) => Wrap(
      children: [
        // Lists of various kinds of widgets that vary quite a lot in size
      ],
    ),
  ),
);
```

My goal is to improve the Scrollbar behaviour without sacrificing any performance at all (so I don't want to increase `cacheExtent` to an arbitrary value for instance).

Ideas:

- As I scroll down for the first time, I intercept the scroll notifications to "replace" them with a better estimate of the `maxScrollExtent`. I can't pre-compute the size of each of my `Wrap`s without rendering them, but I can come up with quite a good estimate of the ratio of the total height they'll need based on their amount of children, for instance.
- When I reach the bottom, I go back to using the default `maxScrollExtent` given by Flutter as the whole ListView was rendered (thus the actual value is better than any estimate I could come up with).
- I always cache the latest value and reuse it as much as possible, as it should be the most accurate. This way, if I scroll back up it won't jump again.

Proof of concept:

```
+ _maxBuiltItemIndex = 0;

+ ({int maxBuiltItemIndex, double maxScrollExtent})? maxScrollExtent;
  Scrollbar(
+   child: NotificationListener<ScrollNotification>(
+     onNotification: (notification) {
+       if (maxScrollExtent == null || maxScrollExtent!.maxBuiltItemIndex < _maxBuiltItemIndex) {
+         maxScrollExtent = _computeMaxScrollExtent(scrollMetrics: notification.metrics); // Custom function to estimate maxScrollExtent or fall back to the default value if _maxBuiltItemIndex == 99
+       }
+       final scrollPosition = PrimaryScrollController.of(context).position;
+       scrollPosition.applyContentDimensions(scrollPosition.minScrollExtent, maxScrollExtent!.maxScrollExtent);
+     },
      child: ListView.builder(
        itemCount: 100,
        itemBuilder: (context, index) {
+        _maxBuiltItemIndex = max(index, _maxBuiltItemIndex);
          return Wrap(
            children: [
              // Lists of various kinds of widgets that vary quite a lot in size
            ],
          );
        },
      ),
+   ),
  );
```

Caveats:

- I don't know what the exact implications of calling `scrollPosition.applyContentDimensions` this way are. I'm not a fan of doing this as it feels like I'm fighting against the framework, and I don't understand the different parts of the scrolling logic in Flutter enough to figure out a clean way to do this.
- I actually had to wrap the `NotificationListener<ScrollNotification>` into a second `NotificationListener<ScrollMetricsNotification>` with a similar logic, otherwise the scrollbar would jump again as the scrolling stops. However, this started having quite an impact on performance, so I also had to come up with some logic to debounce my calls to `scrollPosition.applyContentDimensions` based on the exact nature of each `notification`... Again, I don't understand the scrolling logic enough to confidently deploy this to prod.

That's it for now. Hopefully someone with a better understanding of Flutter can pick it up from there and improve it! I'd love to see a package on pub.dev offering smooth scrolling one day, or even get a future version of Flutter's `ListView`s to take in a new arg like `itemHeightRatioEstimate` so that I don't have to write funky code 😄

**Comment by jiahaog** [comment]
Internal bug: b/452845034

https://dartpad.dev/?id=31b0cfe7dc9b241a347891f4cbbbf3d0 is a minimal repro with just `CustomScrollView` + `SliverList`.
